libs/cpp-ledger/filecoin/impl/ledger_filecoin_impl.cpp

```diff
@@ -0,0 +1,217 @@
+    const SignatureAnswer sign{
+        .r = {signBytes.begin(), signBytes.begin() + 32},
+        .s = {signBytes.begin() + 32, signBytes.begin() + 64},
+        .v = signBytes.at(64),
+        .derSignature = {signBytes.begin() + 65, signBytes.end()}};
+
+    return std::make_tuple(sign, Error{});
+  }
+
+  std::tuple<Bytes, Error> LedgerFilecoinImpl::GetPublicKeySECP256K1(
+      const std::vector<uint32_t> &bip44path) const {
+    Bytes pubkey;
+    Error err;
+
+    std::tie(pubkey, std::ignore, std::ignore, err) =
+        RetrieveAddressPubKeySECP256K1(bip44path, false);
+
+    return std::make_tuple(pubkey, err);
+  }
+
+  std::tuple<Bytes, Bytes, std::string, Error>
+  LedgerFilecoinImpl::GetAddressPubKeySECP256K1(
+      const std::vector<uint32_t> &bip44path) const {
+    return RetrieveAddressPubKeySECP256K1(bip44path, false);
+  }
+
+  std::tuple<Bytes, Bytes, std::string, Error>
+  LedgerFilecoinImpl::ShowAddressPubKeySECP256K1(
+      const std::vector<uint32_t> &bip44path) const {
+    return RetrieveAddressPubKeySECP256K1(bip44path, true);
+  }
+
+  std::tuple<Bytes, Error> LedgerFilecoinImpl::Sign(
+      const std::vector<uint32_t> &bip44path, const Bytes &transaction) const {
+    auto [pathBytes, err] = getBip44bytes(bip44path, kHardenCount);
+
+    if (err != std::nullopt) {
+      return std::make_tuple(Bytes{}, err);
+    }
+
+    const auto chunks = prepareChunks(pathBytes, transaction);
+
+    Bytes response;
+
+    for (size_t chunkId = 0; chunkId < chunks.size(); chunkId++) {
+      const auto &chunk = chunks[chunkId];
+
+      const Byte payloadDesc = (chunkId == 0) ? kPayloadChunkInit
+                                              : (chunkId == chunks.size() - 1)
+                                                    ? kPayloadChunkLast
+                                                    : kPayloadChunkAdd;
+
+      Bytes message{kCLA,
+                    kINSSignSECP256K1,
+                    payloadDesc,
+                    0,
+                    static_cast<Byte>(chunk.size())};
+      message.insert(message.end(), chunk.begin(), chunk.end());
+
+      std::tie(response, err) = device->Exchange(message);
+
+      if (err != std::nullopt) {
+        return std::make_tuple(Bytes{}, err);
+      }
+    }
+
+    return std::make_tuple(response, Error{});
+  }
+
+  std::tuple<Bytes, Bytes, std::string, Error>
+  LedgerFilecoinImpl::RetrieveAddressPubKeySECP256K1(
+      const std::vector<uint32_t> &bip44path, bool requireConfirmation) const {
+    const auto [pathBytes, err] = getBip44bytes(bip44path, kHardenCount);
+
+    if (err != std::nullopt) {
+      return std::make_tuple(Bytes{}, Bytes{}, "", err);
+    }
+
+    const Byte confirm = requireConfirmation ? 1 : 0;
+
+    Bytes message{kCLA,
+                  kINSGetAddrSECP256K1,
+                  confirm,
+                  0,
+                  static_cast<Byte>(pathBytes.size())};
+
+    message.insert(message.end(), pathBytes.begin(), pathBytes.end());
```

Comment:
Maybe .reserve before?

Reply:
It doesn't make sense for only one insertion. There is no difference where the pull of memory is allocated - inside insert() or 1 step before.
It is actual for multiple insertions.